File: python/lokutor/audio_utils.py
"""
Audio utility functions for format conversion, resampling, and PCM processing
"""

import numpy as np
from scipy import signal
from typing import Tuple


def pcm16_to_float32(pcm_data: np.ndarray) -> np.ndarray:
    """
    Convert 16-bit PCM (int16) to 32-bit Float
    
    Args:
        pcm_data: Int16 array of PCM audio
        
    Returns:
        Float32 array normalized to [-1, 1]
    """
    pcm_data = np.asarray(pcm_data, dtype=np.int16)
    return pcm_data.astype(np.float32) / 32768.0


def float32_to_pcm16(float_data: np.ndarray) -> np.ndarray:
    """
    Convert 32-bit Float to 16-bit PCM (int16)
    
    Args:
        float_data: Float32 array normalized to [-1, 1]
        
    Returns:
        Int16 array of PCM audio
    """
    float_data = np.asarray(float_data, dtype=np.float32)
    # Clamp to [-1, 1]
    float_data = np.clip(float_data, -1.0, 1.0)
    # Convert to Int16
    return (float_data * 32767.0).astype(np.int16)


def resample(
    audio: np.ndarray,
    input_rate: int,
    output_rate: int
) -> np.ndarray:
    """
    Resample audio data from one sample rate to another using linear interpolation
    
    Args:
        audio: Float32 array of input audio
        input_rate: Original sample rate in Hz
        output_rate: Target sample rate in Hz
        
    Returns:
        Resampled float32 array
    """
    if input_rate == output_rate:
        return audio.copy()
    
    audio = np.asarray(audio, dtype=np.float32)
    ratio = input_rate / output_rate
    new_length = int(np.round(len(audio) / ratio))
    
    # Use scipy's resampling with high-quality interpolation
    return signal.resample(audio, new_length).astype(np.float32)


def apply_low_pass_filter(
    audio: np.ndarray,
    cutoff_freq: float,
    sample_rate: int,
    order: int = 4
) -> np.ndarray:
    """
    Apply a butterworth low-pass filter for anti-aliasing
    
    Args:
        audio: Float32 array of audio
        cutoff_freq: Cutoff frequency in Hz
        sample_rate: Sample rate in Hz
        order: Filter order (higher = steeper)
        
    Returns:
        Filtered float32 array
    """
    audio = np.asarray(audio, dtype=np.float32)
    
    # Normalize cutoff frequency to Nyquist frequency
    nyquist = sample_rate / 2
    normalized_cutoff = cutoff_freq / nyquist
    
    # Clamp to valid range
    if normalized_cutoff >= 1.0:
        return audio.copy()
    if normalized_cutoff <= 0:
        return np.zeros_like(audio)
    
    # Design butterworth filter
    b, a = signal.butter(order, normalized_cutoff, btype='low')
    
    # Apply filter
    return signal.filtfilt(b, a, audio).astype(np.float32)


def resample_with_anti_aliasing(
    audio: np.ndarray,
    input_rate: int,
    output_rate: int
) -> np.ndarray:
    """
    Resample audio with anti-aliasing low-pass filter
    Best used when downsampling to prevent aliasing artifacts
    
    Args:
        audio: Float32 array of input audio
        input_rate: Original sample rate in Hz
        output_rate: Target sample rate in Hz
        
    Returns:
        Resampled and filtered float32 array
    """
    if input_rate == output_rate:
        return audio.copy()
    
    audio = np.asarray(audio, dtype=np.float32)
    
    # If downsampling, apply low-pass filter first
    if output_rate < input_rate:
        nyquist_freq = output_rate / 2
        cutoff_freq = nyquist_freq * 0.9  # 90% of Nyquist
        audio = apply_low_pass_filter(audio, cutoff_freq, input_rate)
    
    return resample(audio, input_rate, output_rate)


def normalize_audio(audio: np.ndarray, target_peak: float = 0.95) -> np.ndarray:
    """
    Normalize audio amplitude to prevent clipping
    
    Args:
        audio: Float32 array of audio
        target_peak: Peak level to normalize to (0-1)
        
    Returns:
        Normalized float32 array
    """
    audio = np.asarray(audio, dtype=np.float32)
    max_abs = np.max(np.abs(audio))
    
    if max_abs == 0:
        return audio.copy()
    
    scale = target_peak / max_abs
    return (audio * scale).astype(np.float32)


def calculate_rms(audio: np.ndarray) -> float:
    """
    Calculate RMS (Root Mean Square) amplitude
    
    Args:
        audio: Float32 or Int16 array of audio
        
    Returns:
        RMS value (0-1 for float, 0-32768 for int16)
    """
    audio = np.asarray(audio)
    
    if audio.dtype == np.int16:
        # For int16 data
        audio_float = audio.astype(np.float32) / 32768.0
    else:
        # For float data
        audio_float = audio.astype(np.float32)
    
    return float(np.sqrt(np.mean(audio_float ** 2)))


class StreamResampler:
    """
    Stateful resampler for processing audio in chunks
    Useful for streaming audio without buffering entire audio
    """
    
    def __init__(self, input_rate: int, output_rate: int):
        """
        Initialize resampler
        
        Args:
            input_rate: Input sample rate in Hz
            output_rate: Output sample rate in Hz
        """
        self.input_rate = input_rate
        self.output_rate = output_rate
        self.input_buffer = np.array([], dtype=np.float32)
        self.ratio = input_rate / output_rate
    
    def process(self, chunk: np.ndarray, flush: bool = False) -> np.ndarray:
        """
        Process a chunk of audio and return resampled data
        
        Args:
            chunk: Float32 array chunk to process
            flush: If True, output remaining buffered samples
            
        Returns:
            Resampled float32 array (may be empty if more data needed)
        """
        chunk = np.asarray(chunk, dtype=np.float32)
        
        # Add to buffer
        self.input_buffer = np.concatenate([self.input_buffer, chunk])
        
        # Calculate how many output samples we can produce
        if flush:
            output_length = int(np.ceil(len(self.input_buffer) / self.ratio))
            remaining = 0
        else:
            output_length = int(np.floor(len(self.input_buffer) / self.ratio))
            remaining = int(np.ceil(len(self.input_buffer) - output_length * self.ratio))
        
        if output_length == 0:
            if flush:
                self.input_buffer = np.array([], dtype=np.float32)
            return np.array([], dtype=np.float32)
        
        # Resample
        samples_needed = int(output_length * self.ratio)
        to_resample = self.input_buffer[:samples_needed]
        
        if self.input_rate == self.output_rate:
            output = to_resample[:output_length]
        else:
            output = resample(to_resample, self.input_rate, self.output_rate)
            output = output[:output_length]
        
        # Keep remaining in buffer
        self.input_buffer = self.input_buffer[samples_needed:]
        
        return output.astype(np.float32)
    
    def reset(self) -> None:
        """Clear the internal buffer"""
        self.input_buffer = np.array([], dtype=np.float32)


def pcm16_to_bytes(pcm_data: np.ndarray) -> bytes:
    """
    Convert int16 PCM array to raw bytes
    
    Args:
        pcm_data: Int16 array of PCM audio
        
    Returns:
        Raw bytes
    """
    pcm_data = np.asarray(pcm_data, dtype=np.int16)
    return pcm_data.tobytes()


def bytes_to_pcm16(data: bytes) -> np.ndarray:
    """
    Convert raw bytes to int16 PCM array
    
    Args:
        data: Raw bytes
        
    Returns:
        Int16 array of PCM audio
    """
    return np.frombuffer(data, dtype=np.int16)

File: js/src/audio-utils.ts
import { AUDIO_CONFIG } from './types';

/**
 * Audio utility functions for format conversion, resampling, and PCM processing
 */

/**
 * Convert 16-bit PCM (Int16) to 32-bit Float
 * @param int16Data Int16Array of PCM audio
 * @returns Float32Array normalized to [-1, 1]
 */
export function pcm16ToFloat32(int16Data: Int16Array): Float32Array {
  const float32 = new Float32Array(int16Data.length);
  for (let i = 0; i < int16Data.length; i++) {
    float32[i] = int16Data[i] / 32768.0;
  }
  return float32;
}

/**
 * Convert 32-bit Float to 16-bit PCM (Int16)
 * @param float32Data Float32Array normalized to [-1, 1]
 * @returns Int16Array of PCM audio
 */
export function float32ToPcm16(float32Data: Float32Array): Int16Array {
  const int16 = new Int16Array(float32Data.length);
  for (let i = 0; i < float32Data.length; i++) {
    // Clamp to [-1, 1]
    const s = Math.max(-1, Math.min(1, float32Data[i]));
    // Convert to Int16
    int16[i] = s < 0 ? s * 0x8000 : s * 0x7FFF;
  }
  return int16;
}

/**
 * Resample audio data from one sample rate to another using linear interpolation
 * @param input Float32Array of input audio
 * @param inputRate Original sample rate in Hz
 * @param outputRate Target sample rate in Hz
 * @returns Float32Array of resampled audio
 */
export function resample(
  input: Float32Array,
  inputRate: number,
  outputRate: number
): Float32Array {
  if (inputRate === outputRate) {
    return new Float32Array(input);
  }

  const ratio = inputRate / outputRate;
  const outputLength = Math.round(input.length / ratio);
  const output = new Float32Array(outputLength);

  for (let i = 0; i < outputLength; i++) {
    const pos = i * ratio;
    const left = Math.floor(pos);
    const right = Math.min(left + 1, input.length - 1);
    const weight = pos - left;

    // Linear interpolation
    output[i] = input[left] * (1 - weight) + input[right] * weight;
  }

  return output;
}

/**
 * Apply a simple low-pass filter for anti-aliasing during downsampling
 * @param data Float32Array of audio
 * @param cutoffFreq Cutoff frequency in Hz
 * @param sampleRate Sample rate in Hz
 * @returns Filtered Float32Array
 */
export function applyLowPassFilter(
  data: Float32Array,
  cutoffFreq: number,
  sampleRate: number
): Float32Array {
  // Calculate filter coefficient using first-order IIR filter
  const dt = 1 / sampleRate;
  const rc = 1 / (2 * Math.PI * cutoffFreq);
  const alpha = dt / (rc + dt);

  const filtered = new Float32Array(data.length);
  filtered[0] = data[0];

  for (let i = 1; i < data.length; i++) {
    filtered[i] = filtered[i - 1] + alpha * (data[i] - filtered[i - 1]);
  }

  return filtered;
}

/**
 * Resample audio with anti-aliasing low-pass filter
 * Best used when downsampling to prevent aliasing artifacts
 * @param input Float32Array of input audio
 * @param inputRate Original sample rate in Hz
 * @param outputRate Target sample rate in Hz
 * @returns Float32Array of resampled and filtered audio
 */
export function resampleWithAntiAliasing(
  input: Float32Array,
  inputRate: number,
  outputRate: number
): Float32Array {
  if (inputRate === outputRate) {
    return new Float32Array(input);
  }

  // If downsampling, apply low-pass filter first to prevent aliasing
  let processed = input;
  if (outputRate < inputRate) {
    const nyquistFreq = outputRate / 2;
    const cutoffFreq = nyquistFreq * 0.9; // 90% of Nyquist to be safe
    processed = applyLowPassFilter(input, cutoffFreq, inputRate);
  }

  return resample(processed, inputRate, outputRate);
}

/**
 * Convert raw audio samples to Uint8Array (bytes)
 * @param data Int16Array of PCM audio
 * @returns Uint8Array containing PCM bytes
 */
export function pcm16ToBytes(data: Int16Array): Uint8Array {
  return new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
}

/**
 * Convert bytes to Int16Array
 * @param bytes Uint8Array of PCM bytes
 * @returns Int16Array of PCM audio
 */
export function bytesToPcm16(bytes: Uint8Array): Int16Array {
  if (bytes.length % 2 !== 0) {
    // Trim the last byte to avoid RangeError
    bytes = bytes.slice(0, bytes.length - 1);
  }
  return new Int16Array(bytes.buffer, bytes.byteOffset, bytes.length / 2);
}

/**
 * Normalize audio amplitude to prevent clipping
 * @param data Float32Array of audio
 * @param targetPeak Peak level to normalize to (0-1)
 * @returns Normalized Float32Array
 */
export function normalizeAudio(data: Float32Array, targetPeak: number = 0.95): Float32Array {
  let maxAbs = 0;
  for (let i = 0; i < data.length; i++) {
    maxAbs = Math.max(maxAbs, Math.abs(data[i]));
  }

  if (maxAbs === 0) return new Float32Array(data);

  const scale = targetPeak / maxAbs;
  const normalized = new Float32Array(data.length);
  for (let i = 0; i < data.length; i++) {
    normalized[i] = data[i] * scale;
  }
  return normalized;
}

/**
 * Calculate RMS (Root Mean Square) amplitude
 * @param data Float32Array or Uint8Array of audio
 * @returns RMS value (0-1 for normalized float, 0-255 for byte data)
 */
export function calculateRMS(data: Float32Array | Uint8Array): number {
  let sum = 0;
  let length = data.length;

  if (data instanceof Uint8Array) {
    // For byte data (0-128 is silence, 128-255 is audio)
    for (let i = 0; i < length; i++) {
      const v = (data[i] - 128) / 128.0;
      sum += v * v;
    }
  } else {
    // For float data (-1 to 1)
    for (let i = 0; i < length; i++) {
      sum += data[i] * data[i];
    }
  }

  return Math.sqrt(sum / length);
}

/**
 * Create a resample function factory for streaming audio
 * Useful for processing audio in chunks
 */
export class StreamResampler {
  private inputBuffer: Float32Array = new Float32Array(0);
  private inputRate: number;
  private outputRate: number;

  constructor(inputRate: number, outputRate: number) {
    this.inputRate = inputRate;
    this.outputRate = outputRate;
  }

  /**
   * Process a chunk of audio and return resampled data
   * @param inputChunk Float32Array chunk to process
   * @param flush If true, output remaining buffered samples
   * @returns Resampled Float32Array (may be empty if more data needed)
   */
  process(inputChunk: Float32Array, flush: boolean = false): Float32Array {
    // Concatenate new data with buffered data
    const combined = new Float32Array(this.inputBuffer.length + inputChunk.length);
    combined.set(this.inputBuffer);
    combined.set(inputChunk, this.inputBuffer.length);

    const ratio = this.inputRate / this.outputRate;
    let outputLength = Math.floor(combined.length / ratio);

    if (outputLength === 0 && !flush) {
      this.inputBuffer = combined;
      return new Float32Array(0);
    }

    // On flush, output everything including partial frames
    if (flush && outputLength === 0 && combined.length > 0) {
      outputLength = 1;
    }

    const output = new Float32Array(outputLength);

    for (let i = 0; i < outputLength; i++) {
      const pos = i * ratio;
      const left = Math.floor(pos);
      const right = Math.min(left + 1, combined.length - 1);
      const weight = pos - left;

      output[i] = combined[left] * (1 - weight) + combined[right] * weight;
    }

    // Keep remaining samples in buffer
    const consumed = Math.floor(outputLength * ratio);
    this.inputBuffer = combined.slice(consumed);

    return output;
  }

  reset(): void {
    this.inputBuffer = new Float32Array(0);
  }
}
